src/com/facebook/buck/cxx/CxxPrepareForLinkStep.java

```diff
@@ -20,126 +20,65 @@
-import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
 
 /**
  * Prepares argfile for the CxxLinkStep, so all arguments to the linker will be stored in a
  * single file. CxxLinkStep then would pass it to the linker via @path/to/argfile.txt command.
  * This allows us to break the constraints that command line sets for the maximum length of
  * the commands.
  */
-public class CxxPrepareForLinkStep implements Step {
-
-  private final Path argFilePath;
-  private final Path fileListPath;
-  private final ImmutableList<Arg> linkerArgsToSupportFileList;
-  private final Path output;
-  private final ImmutableList<Arg> args;
-  private final Linker linker;
+public class CxxPrepareForLinkStep extends CompositeStep {
 
   private static final Logger LOG = Logger.get(CxxPrepareForLinkStep.class);
 
-  public CxxPrepareForLinkStep(
+  public static CxxPrepareForLinkStep create(
       Path argFilePath,
       Path fileListPath,
       Iterable<Arg> linkerArgsToSupportFileList,
       Path output,
       ImmutableList<Arg> args,
       Linker linker) {
-    this.argFilePath = argFilePath;
-    this.fileListPath = fileListPath;
-    this.linkerArgsToSupportFileList = ImmutableList.copyOf(linkerArgsToSupportFileList);
-    this.output = output;
-    this.args = args;
-    this.linker = linker;
-  }
-
-  private void prepareFileContents() throws IOException {
-    cleanupPreviousFiles();
-    createFileList();
-    createArgFile();
-  }
-
-  private void createFileList() throws IOException {
-    if (Files.notExists(fileListPath.getParent())) {
-      Files.createDirectories(fileListPath.getParent());
-    }
-    if (linkerArgsToSupportFileList.isEmpty()) {
-      LOG.verbose("linkerArgsToSupportFileList are empty, filelist feature is not supported");
-      return;
-    }
-    LOG.verbose("Creating filelist");
-    ImmutableList<String> contents = Arg.stringify(getFileListArgs());
-    storeContentsInFile(contents, fileListPath);
-  }
-
-  private void createArgFile() throws IOException {
-    if (Files.notExists(argFilePath.getParent())) {
-      Files.createDirectories(argFilePath.getParent());
-    }
-    LOG.verbose("Creating argfile");
-    ImmutableList<String> argFileContents =
-        FluentIterable.from(getLinkerOptions())
-            .transform(Javac.ARGFILES_ESCAPER)
-            .toList();
-    storeContentsInFile(argFileContents, argFilePath);
-  }
 
-  private void storeContentsInFile(ImmutableList<String> contents, Path path) throws IOException {
-    try (PrintWriter pr = new PrintWriter(path.toString())) {
-      for (String option : contents) {
-        pr.println(option);
-      }
-    }
-  }
+    ImmutableList<Arg> allArgs = ImmutableList.<Arg>builder()
+        .addAll(StringArg.from(linker.outputArgs(output.toString())))
+        .addAll(args)
+        .addAll(linkerArgsToSupportFileList)
+        .build();
 
-  private void cleanupPreviousFiles() throws IOException {
```

Comment:
You need to maintain this functionality though. Move it into `CxxLink` by adding a new `RmStep` there.